File: tests/test_eigen_tensor.cpp
#include <nanobind/stl/complex.h>
#include <nanobind/eigen/dense.h>
#include <nanobind/eigen/tensor.h>
#include <nanobind/trampoline.h>

namespace nb = nanobind;

using namespace nb::literals;

NB_MODULE(test_eigen_tensor_ext, m) {
    using Tensor3d = Eigen::Tensor<double, 3, 0>;
    using Tensor3i = Eigen::Tensor<int32_t, 3, 0>;
    static_assert(nb::detail::is_eigen_tensor_v<Tensor3d>);
    using RowTensor3d = Eigen::Tensor<double, 3, Eigen::RowMajor>;
    static_assert(nb::detail::make_caster<RowTensor3d>::IsRowMajor);

    using Tensor0d = Eigen::Tensor<double, 0, 0>;

    using MapTensor3d = Eigen::TensorMap<Tensor3d>; // Unaligned by default
    using MapTensor3dConst = Eigen::TensorMap<const Tensor3d>;
    static_assert(nb::detail::is_eigen_tensor_v<MapTensor3d>);
    static_assert(nb::detail::is_eigen_tensor_map_v<MapTensor3d>);

    using RefTensor3d = Eigen::TensorRef<Tensor3d>;
    static_assert(nb::detail::is_eigen_tensor_v<RefTensor3d>);
    static_assert(!nb::detail::is_eigen_tensor_map_v<RefTensor3d>);
    static_assert(nb::detail::is_eigen_tensor_ref_v<RefTensor3d>);

    // -- Plain tensor types

    m.def("add3dTensor", [](const Tensor3d &a, const Tensor3d &b) {
        return a + b;
    }, "a"_a, "b"_a);
    m.def("add3dTensor_nc", [](const Tensor3d &a, const Tensor3d &b) {
        return a + b;
    }, "a"_a.noconvert(), "b"_a.noconvert());
    m.def("square3dTensorR", [](const RowTensor3d &a) {
        return a.square();
    }, "a"_a.noconvert());

    m.def("mul3dTensor", [](double a, const Tensor3d &b) -> Tensor3d {
        return a * b;
    }, "a"_a, "b"_a);

    // -- Refs

    m.def("update3dTensorRef", [](Eigen::TensorRef<Tensor3d> a) {
        a.coeffRef(0, 0, 0) = 42.0;
    }, "a"_a.noconvert());

    // -- Maps - noconvert() is implicit

    m.def("add3dTensorCnstMap", [](MapTensor3dConst a, MapTensor3dConst b) -> Tensor3d {
        return a + b;
    }, "a"_a, "b"_a);

    m.def("castTo3iTensorMap", [](nb::object obj)  {
        return nb::cast<Eigen::TensorMap<Tensor3i, Eigen::Unaligned>>(obj);
    });

    m.def("castTo3iTensorMapCnst", [](nb::object obj)  {
        return nb::cast<Eigen::TensorMap<const Tensor3i, Eigen::Unaligned>>(obj);
    });

    m.def("castTo3iTensorMapAligned", [](nb::object obj)  {
        return nb::cast<Eigen::TensorMap<Tensor3i, Eigen::Aligned>>(obj);
    });

    m.def("castTo0dTensorMap", [](nb::object obj) {
        return nb::cast<Eigen::TensorMap<Tensor0d>>(obj);
    });

    m.def("mul3dTensorMap", [](double a, Eigen::TensorMap<const Tensor3d> b) -> Tensor3d {
        return (a * b).eval();
    }, "a"_a, "b"_a);

    m.def("mul3dTensorMapInPlace", [](double a, Eigen::TensorMap<Tensor3d> b) -> void {
        b = a * b;
    }, "a"_a, "b"_a);

    struct Buffer {
        uint32_t x[18];
        using Tensor3u = Eigen::Tensor<uint32_t, 3>;
        using Map = Eigen::TensorMap<Tensor3u>;

        Map map() { return Map(x, std::array{2, 3, 3}); }
    };

    nb::class_<Buffer>(m, "Buffer")
        .def(nb::init<>())
        .def("map", &Buffer::map, nb::rv_policy::reference_internal);

    struct ClassWithEigenMember {
        Tensor3d member;
        ClassWithEigenMember() : member(2, 1, 2) {
            member.setConstant(1.0);
        }
        const Tensor3d &get_member_ref() { return member; }
        const Tensor3d get_member_copy() { return member; }
    };

    nb::class_<ClassWithEigenMember>(m, "ClassWithEigenMember")
        .def(nb::init<>())
        .def_prop_ro("member_ro_ref", &ClassWithEigenMember::get_member_ref)
        .def_prop_ro("member_ro_copy", &ClassWithEigenMember::get_member_copy)
        .def_rw("member", &ClassWithEigenMember::member);

}

File: tests/test_eigen_tensor.py
import pytest
import pytest

try:
    import numpy as np
    from numpy.testing import assert_array_equal
    import test_eigen_tensor_ext as t
    HAS_NUMPY_AND_EIGEN = True
except ImportError:
    HAS_NUMPY_AND_EIGEN = False

needs_numpy_and_eigen = pytest.mark.skipif(
    not HAS_NUMPY_AND_EIGEN,
    reason="NumPy and Eigen are required")


@needs_numpy_and_eigen
def test01_tensor3d():
    a = np.arange(0, 12, dtype=float).reshape(2, 3, 2)
    assert_array_equal(t.square3dTensorR(a), a ** 2)
    a_colmaj = np.asfortranarray(a)
    with pytest.raises(TypeError, match='incompatible function arguments'):
        t.square3dTensorR(a_colmaj)

    # add3dTensor: col-major inputs (with implicit conversion from row-major)
    b = np.arange(12, 24, dtype=float).reshape(2, 3, 2)
    assert_array_equal(t.add3dTensor(a, b), a + b)
    # row-major numpy arrays are implicitly converted
    assert_array_equal(t.add3dTensor(a, a), a + a)

    # add3dTensor_nc: noconvert — rejects row-major (C-order) arrays
    with pytest.raises(TypeError, match='incompatible function arguments'):
        t.add3dTensor_nc(a, b)
    # Fortran-order (col-major) arrays are accepted without conversion
    assert_array_equal(t.add3dTensor_nc(a_colmaj, np.asfortranarray(b)), a + b)

    # mul3dTensor: scalar * col-major tensor
    assert_array_equal(t.mul3dTensor(2.0, a), 2.0 * a)

    # mul3dTensorMap: read-only TensorMap, returns a new tensor
    assert_array_equal(t.mul3dTensorMap(3.0, a_colmaj), 3.0 * a)

    # mul3dTensorMapInPlace: mutates the array in place via TensorMap
    arr = np.asfortranarray(a.copy())
    t.mul3dTensorMapInPlace(2.0, arr)
    assert_array_equal(arr, 2.0 * a)

    # try non-contiguous
    c = a[:, 1:, :]
    d = a[:, :-1, :]
    assert_array_equal(t.add3dTensor(c, d), c + d)
    with pytest.raises(TypeError, match='incompatible function arguments'):
        t.add3dTensor_nc(c, d)

    with pytest.raises(TypeError, match='incompatible function arguments'):
        t.add3dTensorCnstMap(c, d)

    with pytest.raises(TypeError, match='incompatible function arguments'):
        t.update3dTensorRef(c)


@needs_numpy_and_eigen
def test02_update_tensorref():
    a = np.arange(0, 12, dtype=float).reshape(2, 3, 2)

    arr = np.asfortranarray(a.copy())
    t.update3dTensorRef(arr)
    assert arr[0, 0, 0] == 42.0

    # wrong scalar type is rejected
    with pytest.raises(TypeError, match='incompatible function arguments'):
        t.update3dTensorRef(np.asfortranarray(a).astype(np.int32))

@needs_numpy_and_eigen
def test03_prop():
    for j in range(3):
        c = t.ClassWithEigenMember()
        ref = np.ones((2, 1, 2))
        if j == 0:
            c.member = ref

        for i in range(2):
            member = c.member
            if j == 2 and i == 0:
                member[0, 0, 0] = 10
                ref[0, 0, 0] = 10
            assert_array_equal(member, ref)
            del member

        member = c.member
        assert_array_equal(c.member_ro_ref, ref)
        assert_array_equal(c.member_ro_copy, ref)
        del c
        assert_array_equal(member, ref)

@needs_numpy_and_eigen
def test04_map():
    b = t.Buffer()
    m = b.map()
    for i in range(2):
        for j in range(3):
            for k in range(3):
                m[i, j, k] = i*3*3+j*3+k
    del b
    for i in range(2):
        for j in range(3):
            for k in range(3):
                m[i, j, k] = i*3*3+j*3+k

@needs_numpy_and_eigen
def test05_cast():
    a = np.arange(12, dtype=np.int32).reshape(2, 2, 3, order='F')
    assert_array_equal(t.castTo3iTensorMap(a), a)
    assert_array_equal(t.castTo3iTensorMapAligned(a), a)


@needs_numpy_and_eigen
def test06_zero_size_tensor():
    a = np.ones((0, 2, 4), dtype=np.float64, order='F')
    b = np.ones((0, 2, 4), dtype=np.float64, order='F')
    assert_array_equal(t.add3dTensorCnstMap(a, b), a + b)

    c= np.ones((0, 2, 4), dtype=np.int32, order='F')
    c_map = t.castTo3iTensorMap(c)
    assert_array_equal(c_map, c)
    assert not c_map.flags.owndata
    assert c_map.flags.writeable

    c_map_const = t.castTo3iTensorMapCnst(c)
    assert_array_equal(c_map_const, c)
    assert not c_map_const.flags.owndata
    assert not c_map_const.flags.writeable
    assert_array_equal(t.castTo3iTensorMapAligned(c), c)

    # Pretty much a scalar
    d = np.ones((), order='F')
    d_cast = t.castTo0dTensorMap(d)
    assert_array_equal(d_cast, d)
    assert not d_cast.flags.owndata
